Plugins~/Src/MeshSyncClientBlender/msblenModifiers.cpp

```diff
@@ -0,0 +1,325 @@
+	{
+		if (modifier->type != ModifierType::eModifierType_Nodes) {
+			return;
+		}
+
+		auto nodeModifier = (NodesModifierData*)modifier;
+		auto group = nodeModifier->node_group;
+
+		// Loop through modifier data and get the values:
+		for (auto property : blender::list_range((IDProperty*)nodeModifier->settings.properties->data.group.first)) {
+			if (strstr(property->name, "_use_attribute") || strstr(property->name, "_attribute_name")) {
+				continue;
+			}
+
+			if (doesPropertyUseAttribute(property->name, nodeModifier)) {
+				continue;
+			}
+
+			auto socket = getSocketForProperty(property, group);
+
+			if (socket != nullptr) {
+				auto propertyInfo = ms::PropertyInfo::create();
+
+				switch (property->type) {
+				case IDP_INT: {
+					if (socket->type == SOCK_BOOLEAN) {
+						propertyInfo->set(IDP_Int(property), 0, 1);
+					}
+					else {
+						auto defaultValue = (bNodeSocketValueInt*)socket->default_value;
+						propertyInfo->set(IDP_Int(property), defaultValue->min, defaultValue->max);
+					}
+					break;
+				}
+				case IDP_FLOAT: {
+					auto defaultValue = (bNodeSocketValueFloat*)socket->default_value;
+					propertyInfo->set(IDP_Float(property), defaultValue->min, defaultValue->max);
+					break;
+				}
+				case IDP_DOUBLE: {
+					auto defaultValue = (bNodeSocketValueFloat*)socket->default_value;
+					propertyInfo->set((float)IDP_Double(property), defaultValue->min, defaultValue->max);
+					break;
+				}
+				case IDP_ARRAY: {
+					auto defaultValue = (bNodeSocketValueVector*)socket->default_value;
+					switch (property->subtype) {
+					case IDP_INT: {
+						propertyInfo->set((int*)IDP_Array(property), defaultValue->min, defaultValue->max, property->len);
+						break;
+					}
+					case IDP_FLOAT: {
+						propertyInfo->set((float*)IDP_Array(property), defaultValue->min, defaultValue->max, property->len);
+						break;
+					}
+					}
+					break;
+				}
+				case IDP_STRING: {
+					auto uiData = (bNodeSocketValueString*)property->ui_data;
+					auto val = IDP_String(property);
+					propertyInfo->set(val, strlen(val));
+					break;
+				}
+				default:
+					continue;
+				}
+
+				propertyInfo->path = msblenUtils::get_path(obj);
+				propertyInfo->name = socket->name;
+				propertyInfo->modifierName = modifier->name;
+				propertyInfo->propertyName = property->name;
+				propertyInfo->sourceType = ms::PropertyInfo::SourceType::GEO_NODES;
+				propertyManager->add(propertyInfo);
+			}
+		}
+	}
+
+	void addCustomProperties(const Object* obj, ms::PropertyManager* propertyManager) {
+		if (obj->id.properties) {
+			for (auto property : blender::list_range((IDProperty*)obj->id.properties->data.group.first)) {
+				if (property->ui_data == nullptr && property->type != IDP_STRING) {
+					continue;
+				}
+
+				auto propertyInfo = ms::PropertyInfo::create();
+				switch (property->type) {
+				case IDP_INT: {
```

Comment:
I thought about it but the data structures are different. We could probably make it work with template functions and macros but then it becomes much harder to understand, no?

Reply:
It would be worth trying to reduce duplication. I find it difficult to go through this part of the code and review it properly. There is a lot of nesting too.

Reply:
Like I said, I'm not sure how much more this can be reduced without creating a mess of templates and macros. The data structures are different, we create a PropertyInfo, but one needs the defaults from the node sockets implementation classes for each type, the other uses defaults from the type specific UI classes.